examples/iso_dh/DY.Example.DH.Protocol.Total.Proof.fst

```diff
@@ -0,0 +1,276 @@
+    (exists prin. get_signkey_label vk = principal_label prin /\ (
+      match parse sig_message sig_msg with
+      | Some (SigMsg2 sig_msg2) -> (
+        exists y. sig_msg2.gy == (dh_pk y) /\ event_triggered tr prin (Respond1 sig_msg2.alice prin sig_msg2.gx sig_msg2.gy y)
+      )
+      | Some (SigMsg3 sig_msg3) -> (
+        exists x k. sig_msg3.gx == (dh_pk x) /\ event_triggered tr prin (Initiate2 prin sig_msg3.bob sig_msg3.gx sig_msg3.gy k)
+      )
+      | None -> False
+    ))
+  );
+  sign_pred_later = (fun tr1 tr2 vk msg -> ())
+}
+#pop-options
+
+instance dh_crypto_invs: crypto_invariants = {
+  usages = dh_crypto_usages;
+  preds = dh_crypto_preds;
+}
+
+(*** Proofs ***)
+
+val compute_message1_proof:
+  tr:trace ->
+  alice:principal -> bob:principal -> x:bytes -> si:state_id ->
+  Lemma
+    (requires
+      event_triggered tr alice (Initiate1 alice bob x) /\
+      is_secret (principal_state_label alice si) tr x /\
+      DhKey? (get_usage x)
+    )
+    (ensures 
+      is_publishable tr (compute_message1 alice x)
+    )
+let compute_message1_proof tr alice bob x si =
+  let gx = dh_pk x in
+  assert(is_publishable tr gx);
+  let msg = Msg1 {alice; gx} in
+
+  // This lemma
+  // - requires that msg.gx is is publishable
+  // - ensures that `serialize _ msg` is publishable
+  serialize_wf_lemma message (is_publishable tr) msg;
+
+  // The following code is not needed for the proof.
+  // It just shows what we need to show to prove the lemma.
+  let msgb = compute_message1 alice x in 
+  assert(bytes_invariant tr msgb);
+  assert(get_label msgb `can_flow tr` public);
+  assert(is_publishable tr msgb);
+  ()
+
+val decode_message1_proof:
+  tr:trace ->
+  alice:principal -> bob:principal ->
+  msg_bytes:bytes ->
+  Lemma
+  (requires is_publishable tr msg_bytes)
+  (ensures (
+    match decode_message1 msg_bytes with
+    | Some msg1 -> (
+      is_publishable tr msg1.gx
+    )
+    | None -> True
+  ))
+let decode_message1_proof tr alice bob msg_bytes =
+  match decode_message1 msg_bytes with
+    | Some msg1 -> (
+      // This lemma
+      // - requires that msg_bytes is publishable
+      // - ensures that `msg1.gx` is publishable
+      //   (`msg1` being the result of parsing `msg_bytes` to the type `message1`)
+      parse_wf_lemma message (is_publishable tr) msg_bytes;
+      
+      // The following code is not needed for the proof.
+      // It just shows what we need to show to prove the lemma.
+      assert(bytes_invariant tr msg1.gx);
+      assert(get_label msg1.gx `can_flow tr` public);
+      ()
+    )
+    | None -> ()
+
+val compute_message2_proof:
+  tr:trace -> si:state_id ->
+  alice:principal -> bob:principal ->
+  msg1:message1 -> y:bytes ->
+  sk_b:bytes -> n_sig:bytes ->
+  Lemma
+  (requires (
+    let gy = dh_pk y in
+    event_triggered tr bob (Respond1 alice bob msg1.gx gy y) /\
+    is_publishable tr msg1.gx /\ is_publishable tr gy /\
+    gy == dh_pk y /\
```

Comment:
```suggestion
    event_triggered tr bob (Respond1 alice bob msg1.gx (dh_pk y) y) /\
    is_publishable tr msg1.gx /\
```
Once we remove the trivial facts about `gy` it only appears one time so we can inline its definition in `event_triggered`